src/util/sighash.rs

```diff
@@ -101,6 +101,52 @@ pub struct ScriptPath<'s> {
     leaf_version: LeafVersion,
 }
 
+/// Script path spending details for computing the sighash of a Taproot tx
+pub enum TaprootSpendPath<'s> {
+    /// Full script path, containing the script, leaf version and `OP_CODESEPARATOR` position
+    ScriptPath(ScriptPath<'s>),
+    /// Pre-computed leaf hash and `OP_CODESEPARATOR` position
+    LeafHash {
+        /// The leaf hash
+        leaf_hash: TapLeafHash,
+        /// The `OP_CODESEPARATOR` position
+        code_separator_pos: u32,
+    },
+}
+
+impl<'s> TaprootSpendPath<'s> {
+    /// Return the leaf hash and `OP_CODESEPARATOR` position
+    pub fn leaf_hash_and_code_separator(&self) -> Result<(TapLeafHash, u32), Error> {
+        match self {
+            TaprootSpendPath::ScriptPath(ScriptPath { script, code_separator_pos, leaf_version }) => {
+                let mut enc = TapLeafHash::engine();
+                leaf_version.as_u8().consensus_encode(&mut enc)?;
```

Comment:
Am I right that this will ignore the parity flag and will result in an incorrect `TapLeafHash` for non-even cases?

Reply:
I don't think the parity flag is encoded here, I think it's only in the control block that goes in the witness